crates/sui-graphql-rpc/src/types/cursor.rs

```diff
@@ -169,6 +169,15 @@ impl<C> Page<C> {
         }
     }
 
+    pub(crate) fn with_scan_limit(self, scan_limit: Option<u64>) -> Self {
+        Page {
+            after: self.after,
+            before: self.before,
+            limit: self.limit.min(scan_limit.unwrap_or(self.limit)),
```

Comment:
Reading the rest of the PR it seems like you could just get rid of `with_scan_limit`?

Reply:
For something like
```
{
  "first": 5,
  "scanLimit": 4,
  "filter": {
    "afterCheckpoint": 5,
    "beforeCheckpoint": 20
  }
}
```

iiuc we would expect 4 results, and hasNextPage = true if there is at least one more result. What we get back from graphql, without setting the limit, would be 5 results. This is because scanLimit sets the inclusive upper bound, and since `paginate_query` sets `limit + 2`, we'll end up picking up the 5th result in the set.

For some `scanLimit`, it would be some starting point inclusive of `scanLimit`, for example, starting at 0 with scanLimit of 4, the last tx_sequence_number that could be in the result would be 3; consequently, i first tried setting limit = limit - 1, but then the following filters would become an edge case:
```
{
  "first": 1,
  "after": "eyJjIjoyNjAzNDgyLCJ0Ijo2fQ",
  "scanLimit": 1,
  "filter": {
    "afterCheckpoint": 5,
    "beforeCheckpoint": 20
  }
}
``` 
where we know there are more nodes but there should also be a previous page, and we don't render the nodes - this is because scanLimit sets the difference to limit - 1, and paginate_query will fetch limit + 2, and see that "If after trimming, we're going to return no elements" - basically we can't advance the cursor.

At least behind the scenes, I think we need to set the range so that the beginning and end are at least 2 elements apart

Reply:
but i think i understand what you mean. say i have a combination of filters that would yield 5 results, each 10 million apart. If my limit is 1 and my scanLimit is 10 million, then i would expect a total of 5 pages each with a single element. Conversely, a limit of 10 million and a scanLimit of 1 would mean a lot of pages to go through